### PR TITLE
Spend money reminder

### PR DESCRIPTION
**Pull request by roasher**
closes https://github.com/MisterRnobe/sodexo-tg-bot/issues/26

### CODE REVIEW

File: src/main/java/ru/nmedvedev/model/UserDb.java [hunk 1]
```diff
@@ -2,84 +2,34 @@
...
 
     @Deprecated(forRemoval = true)
     public static UserDb withId(ObjectId id,
                                 Long chatId,
                                 String card,
-                                Boolean subscribed) {
-        var userDb = new UserDb(chatId, card, subscribed, null);
+                                Boolean subscribed,
+                                Boolean subscribedToSpendMoneyReminder) {
+        var userDb = new UserDb(id, chatId, card, subscribed, subscribedToSpendMoneyReminder, null);
```
Comment (MisterRnobe) [inline src/main/java/ru/nmedvedev/model/UserDb.java:31]: This method is deprecated. Let's remove the if it is not used

File: src/main/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderBusinessLogicGateService.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+@ApplicationScoped
+public class SpendMoneyReminderBusinessLogicGateService {
+
+    public boolean needToSendNotification(ReminderDayEnum day, Double amount) {
+        switch (day) {
+            case NOT_A_DAY_FOR_A_REMINDER: return false;
+            case HALF_MONTH: return amount >= 1500;
+            case LAST_WORKING_DAY_MINUS_FOUR: return amount >= 1000;
+            case LAST_WORKING_DAY_MINUS_TREE: return amount >= 500;
+            case LAST_WORKING_DAY_MINUS_ONE: return amount >= 400;
```
Comment (MisterRnobe) [inline src/main/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderBusinessLogicGateService.java:14]: Maximum balance may change one day so these values should be modified. I suggest you to move the to the app config.

File: src/test/java/ru/nmedvedev/Helper.java [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 
 public final class Helper {
 
     public static final String CARD = String.valueOf(Math.abs(new Random().nextLong()));
     public static final Long CHAT = new Random().nextLong();
 
+    static {
+        Locale.setDefault(Locale.ROOT);
+    }
+
```
Comment (MisterRnobe) [inline src/test/java/ru/nmedvedev/Helper.java:14]: What does this change do?

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import static org.hamcrest.Matchers.is;
+
+class RemindDayProviderServiceTest {
+
+    private RemindDayProviderService remindDayProviderService = new RemindDayProviderService();
+
+    @Test
+    void returnHalfOfTheMonth() {
+        assertThat(remindDayProviderService.getDay(LocalDate.of(2020, 1, 15)), is(ReminderDayEnum.HALF_MONTH));
+        assertThat(remindDayProviderService.getDay(LocalDate.of(2020, 1, 14)), is(ReminderDayEnum.NOT_A_DAY_FOR_A_REMINDER));
```
Comment (MisterRnobe) [inline src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java:17]: I guess one test method shoud contain only one target method invocation. You can vary arguments using [parameterized tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests)

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        when(remindDayProviderService.getDay(any())).thenReturn(ReminderDayEnum.NOT_A_DAY_FOR_A_REMINDER);
+        service.sendReminders();
+        verifyNoInteractions(spendMoneyReminderBusinessLogicGateService);
+        verifyNoInteractions(userRepository);
+        verifyNoInteractions(telegramService);
+        verifyNoInteractions(replyButtonsProvider);
+    }
+
+    @Test
+    void doNotSendIfUserHasNoAmount() {
```
Comment (MisterRnobe) [inline src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java:50]: Did you mean `NoCard`?
```suggestion
    void doNotSendIfUserHasNoCard() {
```

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        verifyNoInteractions(telegramService);
+        verifyNoInteractions(spendMoneyReminderBusinessLogicGateService);
+    }
+
+    @Test
+    void doNotSendIfUserFailedBusinessLogicGate() {
+        when(remindDayProviderService.getDay(any())).thenReturn(ReminderDayEnum.LAST_WORKING_DAY_MINUS_ONE);
+        when(userRepository.findSubscribedToSpendMoneyReminderWithCardAndChat())
+                .thenReturn(Multi.createFrom().items(
+                        UserDb.builder().latestOperation(HistoryDb.builder().amount(100.).build()).build()
```
Comment (MisterRnobe) [inline src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java:67]: HistoryDb's `amount` field is **not** current user's balance. This field represents how much money were credited to a card or debited from a card. It looks like you should check user's balance from [sodexo api](https://github.com/MisterRnobe/sodexo-tg-bot/blob/master/src/main/java/ru/nmedvedev/rest/SodexoClient.java#L22)

File: src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package ru.nmedvedev.service.spendmoneyreminder;
+
+public enum ReminderDayEnum {
+    NOT_A_DAY_FOR_A_REMINDER(""),
+    HALF_MONTH("Дружок. Половина месяца прошло. У тебя осталось %.2f руб. Не забудь потратить)"),
+    LAST_WORKING_DAY_MINUS_FOUR("На твоём счету %.2f руб. Через 3 дня они сгорят."),
+    LAST_WORKING_DAY_MINUS_TREE("Предпоследний день, чтобы потратить свои %.2f руб."),
+    LAST_WORKING_DAY_MINUS_ONE("Ахтунг, комрад! У тебя ещё %.2f руб. на карте, а сегодня последний день чтобы их потратить!!!");
```
Comment (MisterRnobe) [inline src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java:8]: These texts will definitely encourage me to spend money asap. I think adding emojis will make them more emotional

File: src/test/java/ru/nmedvedev/Helper.java (hunk 3 above)
Comment (roasher) [reply src/test/java/ru/nmedvedev/Helper.java:14]: That makes tests passed on Windows

File: src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java (hunk 7 above)
Comment (roasher) [reply src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java:8]: I'm happy with text only. If you want me to do add something additional, please provide some analytics.

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java (hunk 4 above)
Comment (roasher) [reply src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java:17]: I can vary only one argument as far as I know. In tests here I need to vary at least 3.

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java (hunk 6 above)
Comment (roasher) [reply src/test/java/ru/nmedvedev/service/spendmoneyreminder/SpendMoneyReminderServiceTest.java]: ок

File: src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java (hunk 4 above)
Comment (MisterRnobe) [reply src/test/java/ru/nmedvedev/service/spendmoneyreminder/RemindDayProviderServiceTest.java:17]: No, you can vary as many arguments as you wish

File: src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java (hunk 7 above)
Comment (MisterRnobe) [reply src/main/java/ru/nmedvedev/service/spendmoneyreminder/ReminderDayEnum.java:8]: Ok, let's start with these ones then

File: src/main/java/ru/nmedvedev/rest/SodexoCustomClient.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import io.smallrye.mutiny.Uni;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import ru.nmedvedev.model.Balance;
+import ru.nmedvedev.model.SodexoData;
+import ru.nmedvedev.model.SodexoResponse;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class SodexoCustomClient {
```
Comment (MisterRnobe) [inline src/main/java/ru/nmedvedev/rest/SodexoCustomClient.java:12]: Seems like this class has only `getAmount` additional method. Couldn't it be moved to `SodexoClient` and be implemented as `default`?

Comment (roasher) [reply src/main/java/ru/nmedvedev/rest/SodexoCustomClient.java:12]: That was my first attempt but it failed. Don't know why - there're simply no request. Can't find any examples with microprofile and default method on the net

Comment (MisterRnobe) [reply src/main/java/ru/nmedvedev/rest/SodexoCustomClient.java:12]: Alright, let it be so

**Review by MisterRnobe — APPROVED** [review]
Thanks for enhacement!

**Comment by roasher** [comment]
U welcome

**Pull request merged**